### ISSUE TITLE
semi rule "never" warns against necessary semicolons

### ISSUE THREAD

**Issue by yyx990803** [eslint/eslint#1580]
ESLint version: 0.10.2

To reproduce, run the following in Node:

``` js
var linter = require('eslint').linter
var msgs = linter.verify('a++; b++', {
  rules: {
    semi: [2, 'never']
  }
})
console.log(msgs)
```

ESLint complains about extra semicolon at line 1, column 4 - however if we remove the semicolon, the code becomes invalid and will throw SyntaxError: Unexpected identifier when executed.

The root cause is that ASI only works when a `\n` is present, however the `semi` rule is not checking whether the next statement is starting on a newline when giving the extra semicolon warning.
